pkg/controllers/vdb/replicatedupgrade_reconciler.go

```diff
@@ -93,42 +101,192 @@ func (r *ReplicatedUpgradeReconciler) Reconcile(ctx context.Context, _ *ctrl.Req
+// update any necessary objects the upgrade depends on.
+func (r *ReplicatedUpgradeReconciler) runObjReconciler(ctx context.Context) (ctrl.Result, error) {
+	rec := MakeObjReconciler(r.VRec, r.Log, r.VDB, r.PFacts, ObjReconcileModeAll)
+	return rec.Reconcile(ctx, &ctrl.Request{})
+}
+
+// runAddSubclusterReconciler will run the reconciler to create any necessary subclusters
+func (r *ReplicatedUpgradeReconciler) runAddSubclusterReconciler(ctx context.Context) (ctrl.Result, error) {
+	rec := MakeDBAddSubclusterReconciler(r.VRec, r.Log, r.VDB, r.PFacts.PRunner, r.PFacts, r.Dispatcher)
+	return rec.Reconcile(ctx, &ctrl.Request{})
+}
+
+// runAddNodesReconciler will run the reconciler to scale out any subclusters.
+func (r *ReplicatedUpgradeReconciler) runAddNodesReconciler(ctx context.Context) (ctrl.Result, error) {
+	rec := MakeDBAddNodeReconciler(r.VRec, r.Log, r.VDB, r.PFacts.PRunner, r.PFacts, r.Dispatcher)
+	return rec.Reconcile(ctx, &ctrl.Request{})
+}
+
+// generatePrimariesForReplicaGroupB will create new secondary subclusters for each of the
+// primaries that exist. This is a pre-step to setting up replica group B, which
+// will eventually exist in its own sandbox.
+func (r *ReplicatedUpgradeReconciler) generatePrimariesForReplicaGroupB(ctx context.Context) (ctrl.Result, error) {
+	// Early out if subclusters have already been assigned to replica group B.
+	if r.countSubclustersForReplicaGroup(vmeta.ReplicaGroupBValue) != 0 {
+		return ctrl.Result{}, nil
 	}
 
-	// Get the subcluster statefulsets. We sort this list so our algorithm for
-	// replica group assignment is consistent.
-	stss, err := r.Manager.Finder.FindStatefulSets(ctx, iter.FindExisting|iter.FindSorted)
+	updated, err := r.Manager.updateVDBWithRetry(ctx, r.addNewSubclustersForPrimaries)
 	if err != nil {
-		return ctrl.Result{}, err
+		return ctrl.Result{}, fmt.Errorf("failed trying to update VDB with new subclusters: %w", err)
+	}
+	if updated {
+		r.Log.Info("new secondary subclusters added to mimic the primaries", "len(subclusters)", len(r.VDB.Spec.Subclusters))
+	}
+	return ctrl.Result{}, nil
+}
+
+// sandboxReplicaGroupB will move all of the subclusters in replica B to a new sandbox
+func (r *ReplicatedUpgradeReconciler) sandboxReplicaGroupB(ctx context.Context) (ctrl.Result, error) {
+	return ctrl.Result{}, errors.New("sandbox of replica group B is not yet implemented")
+}
+
+// addNewSubclustersForPrimaries will come up with a list of subclusters we
+// need to add to the VerticaDB to mimic the primaries. The new subclusters will
+// be added directly to r.VDB.
+func (r *ReplicatedUpgradeReconciler) addNewSubclustersForPrimaries() (bool, error) {
+	oldImage, found := r.Manager.fetchOldImage()
+	if !found {
+		return false, errors.New("Could not find old image needed for new subclusters")
 	}
+	newSubclusters := []vapi.Subcluster{}
+	scMap := r.VDB.GenSubclusterMap()
+	for i := range r.VDB.Spec.Subclusters {
```

Comment:
Anyway the code inside this for block can be turned into a function?

Reply:
Yes, will do. Good observation.